### ISSUE TITLE
'allowLocalMutation' flag in [immutable-data] rule

### ISSUE THREAD

**Issue by snowinmars** [jonaskello/eslint-plugin-functional#201]
# Suggestion
Allow local objects and arrays to be mutable.

# Example
```typescript
ngOnChanges(changes: NgChanges<MyComponent>): void {
  const nodes: Node<Data>[] = [];

  // changes.tracks.currentValue is Map<Id, Data>
  changes.tracks.currentValue.forEach((id, data) => {
    nodes.push({ // <- error    Modifying an array is not allowed       functional/immutable-data
      ...data,
    });
  });

  this.nodes = nodes;
}
```

# Reason
I'm working on a Angular project. I'm not sure, do I have enough faith to write the project using functional style. But state mutation is evil, so I want to check it with the `eslint-plugin-functional`. I want to add FP as gentle, as I can; step by step.

The example above can be fixed in three ways.
- I can rewrite it in functional style - change `Map` with some lib, change `forEach`, etc. It's a good thing to do, but I can't rewrite all my code in one night.
- I can rewrite it, f.e. as below, but my colleagues will be asking for perfomance affection; and I'm not sure what to say here. And there are a lot of places where local items mutate. Each case is different, each case should be fixed in it's own way.
```typescript
  changes.tracks.currentValue.forEach((id, data) => {
    nodes = [...nodes, data]
  });
```
- I could've say: 'hey, that's a local item, forget about it for now' - but I can't. There's no `allowLocalMutation` flag in [immutable-data] rule.

Is it a good idea to allow cases like this?

**Comment by RebeccaStevens** [comment]
`allowLocalMutation` can't really be added to this rule because we can't really determine what is "local mutation". In other rules "local mutation" is considered anything that isn't top level but that wouldn't work for this rule as this wouldn't line up with what most users would expect as soon as nested functions come into play.

I'd recommend using the `ignoreAccessorPattern` option for your case. If you set it to something like this `{ "ignoreAccessorPattern": "**.mutable_*" }` then you could rewrite your example function as so:

```ts
ngOnChanges(changes: NgChanges<MyComponent>): void {
  const mutable_nodes: Node<Data>[] = [];

  // changes.tracks.currentValue is Map<Id, Data>
  changes.tracks.currentValue.forEach((id, data) => {
    mutable_nodes.push({
      ...data,
    });
  });

  this.nodes = mutable_nodes;
}
```
